### PR TITLE
GH-36217: [MATLAB]  Add arrow.array.TimestampArray

### LINKED ISSUE #36217: [MATLAB] Add `arrow.array.TimestampArray`
### Describe the enhancement requested

Add a MATLAB class to represent timestamp data. `Timestamp` would map to the MATLAB `datetime` type when calling `toMATLAB`

### Component(s)

MATLAB

### PR DESCRIPTION
**Pull request by sgilmore10**
### Rationale for this change

We would like to extend the MATLAB interface to support timestamp data.

### What changes are included in this PR?

1. Added a new `arrow.array.MATLABArray` class that can be converted to/from a MATLAB `datetime` array.
2. Added a new type class called `arrow.type.TimestampType`. 
3. Added a new enum  class called `arrow.type.TimeUnit`.
 
**Example**

```matlab
>> dates = datetime(2023, 6, 27) + days(0:3)'

dates = 

  3×1 datetime array

   27-Jun-2023
   28-Jun-2023
   29-Jun-2023

>> timestampArray = arrow.array.TimestampArray(dates)

timestampArray = 

[
  2023-06-27 00:00:00.000000,
  2023-06-28 00:00:00.000000,
  2023-06-29 00:00:00.000000
]

>> fromArrow = datetime(timestampArray)

fromArrow = 

  3×1 datetime array

   27-Jun-2023
   28-Jun-2023
   29-Jun-2023

>> isequal(dates, fromArrow)

ans =

  logical

   1
``` 

`TimestampArray` uses `Microsecond` as the default `TimeUnit.` However, you can specify the `TimeUnit` via a name-value pair during construction.

```matlab

>> timestampArray = arrow.array.TimestampArray(dates, TimeUnit="second")

timestampArray = 

[
  2023-06-27 00:00:00,
  2023-06-28 00:00:00,
  2023-06-29 00:00:00
]

>> timestampArray.Type

ans = 

  TimestampType with properties:

      TimeZone: ""
      TimeUnit: Second
            ID: Timestamp
      BitWidth: 64
     NumFields: 0
    NumBuffers: 2
```

The `TimestampArray` is timezone-aware if the MATLAB `datetime` array's TimeZone property is set:

```matlab
>> dates = datetime(2023, 6, 27, TimeZone="America/Anchorage") + days(0:3)';
>> timestampArray = arrow.array.TimestampArray(dates);
>> timestampArray.Type.TimeZone

ans = 

    "America/Anchorage"
```

Lastly, `arrow.array.TimestampArray` treats `NaT` as null values by default. However, users can control this behavior by supplying either `InferNulls` or `Valid` as name-value pairs.

```matlab
>> dates = [datetime(2023, 6, 27) NaT datetime(2023, 6, 29)];
>> timestampArray = arrow.array.TimestampArray(dates)

timestampArray = 

[
  2023-06-27 00:00:00.000000,
  null,
  2023-06-29 00:00:00.000000
]

>> timestampArray = arrow.array.TimestampArray(dates, Valid=3)

timestampArray = 

[
  null,
  null,
  2023-06-29 00:00:00.000000
]
```

### Are these changes tested?

Added three new test files:

1. `tTimestampArray.m`
2.`tTimestampType.m`
3. `tTimeUnit.m`

### Are there any user-facing changes?
Yes.

1. Added `arrow.array.TimestampArray`

### Notes

Thank you @kevingurney for the help!
* Closes: #36217

### CODE REVIEW

File: matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+#include "arrow/matlab/type/time_unit.h"
+#include "arrow/util/utf8.h"
+#include "arrow/type.h"
+#include "arrow/builder.h"
+
+
+namespace arrow::matlab::array::proxy {
+
+    namespace {
+        const uint8_t* getUnpackedValidityBitmap(const ::matlab::data::TypedArray<bool>& valid_elements) {
```
Comment (kou) [inline matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc:33]: Can we unify this and a similar function in `numeric_array.h`?

**Review by kou — APPROVED** [review]
+1

File: matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc (hunk 1 above)
Comment (sgilmore10) [reply matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc:33]: Definitely! That's a good idea.

Comment (sgilmore10) [reply matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc:33]: Actually, it looks like @kevingurney has already done this. He's currently working on adding an `arrow.array.StringArray` class and created a shared utility called [unpacked_as_ptr](https://github.com/mathworks/arrow/blob/cd4c7cc36e2d21894399387fb51ceda9cf86d532/matlab/src/cpp/arrow/matlab/bit/unpack.cc#L42).  Is it alright if I create an issue to use that utility in `timestamp_array.cc` once Kevin's changes are merged?

Comment (kou) [reply matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc:33]: OK. I merge this.

Comment (sgilmore10) [reply matlab/src/cpp/arrow/matlab/array/proxy/timestamp_array.cc:33]: Thanks @kou! I've made an [issue](https://github.com/apache/arrow/issues/36335) to track this.

**Pull request merged**